enhancements/network/ovn-k-localnet-api.md

```diff
@@ -0,0 +1,507 @@
+metadata:
+  name: test-net
+  namespace: red
+finalizers:
+ - k8s.ovn.org/user-defined-network-protection
+labels:
+  k8s.ovn.org/user-defined-network: ""
+ownerReferences:
+- apiVersion: k8s.ovn.org/v1
+  blockOwnerDeletion: true
+  controller: true
+  kind: ClusterUserDefinedNetwork
+  name: test-net
+spec:
+    config: > '{
+        "cniVersion": "1.0.0",
+        "type": "ovn-k8s-cni-overlay"
+        "netAttachDefName": "blue/test-net",
+        "role": "secondary",
+        "topology": "localnet",
+        "name": "cluster.udn.test-net",
+        "physicalNetworkName: "tenantblue",
+        "subnets": "192.168.0.0/16,2001:dbb::/64",
+        "allowPersistentIPs": true,
+        "excludeSubents: "10.100.50.0/24",
+        "mtu": 9000,
+        "vlanID": 200
+    }'
+```
+
+#### Go types
+The CUDN `spec.network.topology` field should be extended to accept `Localnet` string.
+And the `spec.network` struct `NetworkSpec` should be to have additional field for the localnet topology  configuration:
+```go
+const NetworkTopologyLocalnet NetworkTopology = "Localnet"
+
+...
+
+// NetworkSpec defines the desired state of UserDefinedNetworkSpec.
+// +union
+type NetworkSpec struct {
+    // Topology describes network configuration.
+    //
+    // Allowed values are "Layer3", "Layer2" and "Localnet".
+    // Layer3 topology creates a layer 2 segment per node, each with a different subnet. Layer 3 routing is used to interconnect node subnets.
+    // Layer2 topology creates one logical switch shared by all nodes.
+    // Localnet topology attach to the overlay local network. Enables egress to the provider's physical network. 
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    // +unionDiscriminator
+    Topology NetworkTopology `json:"topology"`
+    
+    ...
+    
+    // Localnet is the Localnet topology configuration.
+    // +optional
+    Localnet *LocalnetConfig `json:"localnet,omitempty"`
+}
+```
+
+The CUDN spec should have additional validation rule for `spec.network.topology` field:
+```go
+// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkSpec struct {
+    ...
+    // +required
+    Network NetworkSpec `json:"network"`
+}
+```
+
+As of today the CUDN has validation to ensure spec.network is immutable:
+```go
+// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Network spec is immutable"
+```
+This validation should be changed to allow mutation localnet topology configurations, at least MTU and VLAN.
+
+Localnet topology configuration type 
+```go
+type LocalnetConfig struct {
+    // Role describes the network role in the pod.
+    //
+    // Allowed value is "Secondary".
+    // Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    Role NetworkRole `json:"role"`
```

Comment:
Is the `NetworkRole` a type that already exists today? Does it have the appropriate enum tags to limit the allowed values to only `Secondary`?

Reply:
Are you intending for this field to also allow setting `Primary` and `Secondary`?

Reply:
Clarifying here: there is no use case for having a primary UDN whose topology is `localnet`.

@everettraven what does the openshift API guidelines say about this scenario ?... What is more important ? Consistency across the API (forcing the user to specify *always* the same value) or make the user provide less input ?

Reply:
My specific concern here is that you specify that a user can only set `Secondary` in the Godoc, but the type you are using has validation that allows setting both `Primary` and `Secondary`. There is no additional validation that I see enforces that this must only be set to `Secondary`. Without that explicit validation in place your Godoc is not accurate.

Reply:
I added additional validation to ensure only Secondary value is accepted, 
please see https://github.com/openshift/enhancements/pull/1750/files#diff-65ff4f11b53425ec17e6fbd228843c58470c309f5031c38c345010775bb378d9R432